docs/cloud-integration.md

```diff
@@ -0,0 +1,953 @@
+    Recommended S3 client for Spark releases built on Apache Hadoop 2.7 or later.
+    </td>
+  </tr>
+  <tr>
+    <td><code>s3n://</code></td>
+    <td>Amazon S3</a>
+    <td>
+    Deprected S3 client; only use for Spark releases built on Apache Hadoop 2.6 or earlier.
+    </td>
+  </tr>
+  <tr>
+    <td><code>s3://</code></td>
+    <td>Amazon S3 on Amazon EMR</a>
+    <td>
+    Amazon's own S3 client; use only and exclusivley in Amazon EMR.
+    </td>
+  </tr>
+  <tr>
+    <td><code>wasb://</code></td>
+    <td>Azure Storage</a>
+    <td>
+    Client for Microsoft Azure Storage; since Hadoop 2.7.
+    </td>
+  </tr>
+  <tr>
+    <td><code>swift://</code></td>
+    <td>OpenStack Swift</a>
+    <td>
+    Client for OpenStack Swift object stores.
+    </td>
+  </tr>
+  <tr>
+    <td><code>gs://</code></td>
+    <td>Google Cloud Storage</a>
+    <td>
+    Google's client for their cloud object store.
+    </td>
+  </tr>
+</table>
+
+
+### <a name="working_with_amazon_s3"></a>Working with Amazon S3
+
+Amazon's S3 object store is probably the most widely used object store —it is also the one
+with the most client libraries. This is due to the evolution of Hadoop's support, and Amazon
+offering Hadoop and Spark as its EMR service, along with its own S3 client.
+
+The recommendations for which client to use depend upon the version of Hadoop on the Spark classpath.
+
+<table class="table">
+  <tr><th>Hadoop Library Version</th><th>Client</th></tr>
+  <tr>
+    <td>Hadoop 2.7+ and commercial products based on it</a>
+    <td><code>s3a://</code></td>
+  </tr>
+  <tr>
+    <td>Hadoop 2.6 or earlier</a>
+    <td><code>s3n://</code></td>
+  </tr>
+  <tr>
+    <td>Amazon EMR</a>
+    <td><code>s3://</code></td>
+  </tr>
+</table>
+
+Authentication is generally via properties set in the spark context or, in YARN clusters,
+`core-site.xml`.
+Versions of the S3A client also support short-lived session credentials and IAM authentication to
+automatically pick up credentials on EC2 deployments. Consult the appropriate Hadoop documentation for specifics.
+
+`spark-submit` will automatically pick up and propagate `AWS_ACCESS_KEY` and `AWS_SECRET_KEY`
+from the environment variables set in the environment of the user running `spark-submit`; these
+will override any set in the configuration files.
+
+Be aware that while S3 buckets support complex access control declarations, Spark needs
+full read/write access to any bucket to which it must write data. That is: it does not support writing
+to buckets where the root paths are read only, or not readable at all.
+
+#### <a name="s3a"></a>S3A Filesystem Client: `s3a://`
+
+The ["S3A" filesystem client](https://hadoop.apache.org/docs/stable2/hadoop-aws/tools/hadoop-aws/index.html)
+shipped with in Hadoop 2.6, and has been considered ready for production use since Hadoop 2.7.1
+
+*The S3A connector is the sole S3 connector undergoing active maintenance at the Apache, and
+should be used wherever possible.*
+
+**Classpath**
```

Comment:
While all of the info here is useful and correct I assume, some of these sections seem to be relevant to developers only and not end users. What of this do I need to know as an end user? do I need to put this stuff on my classpath? It could be a dumb question but I thought part of the idea was that this was being pulled in automatically, or had to be.

I think the same applies to a lot of asides or back-story here. Some is good but the focus here has to be what it is Spark users need to know.